19483 Port Test Unit/minitest tests to RSpec

myabc opened the pull request:
Rebased, updated. Succeeds #2115.
## What this PR/WP does not do:
- update Test-Unit/xUnit-style assertions. A separate PR will handle this (indeed I've been working on a [tool to help convert these assertions automatically](https://github.com/myabc/unassertive)).
- remove the `rake test` tasks: we should hold off doing this until all CI jobs are updated.
## What this PR/WP does:
- rewrites Test-Unit tests using RSpec DSL.
- moves Test-Units into a `spec/legacy` folder.
- updates Rake tasks: 
  there are two separate `rake spec:core` and `rake spec:legacy` tasks, which are in turn, dependencies of `rake spec`.

The legacy specs are generally not good about cleaning up after themselves: for example, leaving state in the database. As such, the following will still produce errors:

``` bash
rspec spec/**/*_spec.rb
rspec spec/legacy/**/*_spec.rb
```

while this will work:

``` bash
rspec spec/legacy/unit/**/*_spec.rb
rspec spec/legacy/functional/**/*_spec.rb
rspec spec/legacy/integration/**/*_spec.rb
rspec spec/models/**/*_spec.rb
```

The Rake tasks take this into account. 

https://community.openproject.org/work_packages/19483

myabc committed:
First spike porting Test::Unit tests to RSpec

* Convert tests from Test::Unit to RSpec syntax, but keep Test::Unit-
  style assertions, configuring RSpec to support `:minitest` assertions
* Update Helpers and backported Shoulda macros to work.

Tests can be run with RSpec as follows:

    be rspec -Itest test/**/*_test.rb

Or by Suite:

    be rspec -Itest test/unit/**/*_test.rb
    be rspec -Itest test/functional/**/*_test.rb
    be rspec -Itest test/integration/**/*_test.rb

Tests should also be executable with the Rails Rake test tasks.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove never implemented performance test

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
HACK: Reset I18n.locale to en before each spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
No query params w/X-OpenProject-API-Key header

myabc committed:
Bump shoulda-matchers, specify shoulda-context

Require shoulda-matchers explicitly in `spec_helper.rb`
(as recommended in README:
https://github.com/thoughtbot/shoulda-matchers/tree/v2.8.0#rspec)

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove REPOSITORY_PATH constants from repo specs

* Use more idiomatic RSpec `skip` for skipping examples.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Ensure Repository unit spec uses correct settings

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Ensure Journal unit spec uses correct settings

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Ensure integration specs use correct login setting

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Comment failing assertion in IssueNestedSet test

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Replace deprecated shoulda #set_the_flash in specs

Signed-off-by: Alex Coles <alex@alexbcoles.com>

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://myabc-feature-remove-test-unit.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://myabc-feature-remove-test-unit.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

myabc committed:
Remove spec controller boilerplate; Replace ivars

Prefer methods provided by rspec-rails.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Move legacy specs under spec/legacy

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Replace more deprecated shoulda matchers in specs

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Ensure Changeset legacy spec uses correct settings

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove Test::Unit suite from Travis build matrix

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Convert specs to RSpec 3.2.0 syntax with Transpec

This conversion is done by Transpec 3.1.0 with the following command:
    transpec -f -c "bundle exec rspec -Itest" test/functional/account_controller_test.rb test/functional/activities_controller_test.rb test/functional/admin_controller_test.rb test/functional/application_controller_test.rb test/functional/attachments_controller_test.rb test/functional/boards_controller_test.rb test/functional/custom_fields_controller_test.rb test/functional/enumerations_controller_test.rb test/functional/groups_controller_test.rb test/functional/help_controller_test.rb test/functional/journals_controller_test.rb test/functional/mail_handler_controller_test.rb test/functional/messages_controller_test.rb test/functional/my_controller_test.rb test/functional/project_enumerations_controller_test.rb test/functional/projects_controller_test.rb test/functional/repositories_controller_test.rb test/functional/repositories_filesystem_controller_test.rb test/functional/repositories_git_controller_test.rb test/functional/repositories_subversion_controller_test.rb test/functional/roles_controller_test.rb test/functional/search_controller_test.rb test/functional/settings_controller_test.rb test/functional/sys_controller_test.rb test/functional/time_entries/reports_controller_test.rb test/functional/timelog_controller_test.rb test/functional/types_controller_test.rb test/functional/user_mailer_test.rb test/functional/users_controller_test.rb test/functional/watchers_controller_test.rb test/functional/welcome_controller_test.rb test/functional/wiki_controller_test.rb test/functional/wikis_controller_test.rb test/functional/workflows_controller_test.rb test/integration/api_test/disabled_rest_api_test.rb test/integration/api_test/http_accept_auth_test.rb test/integration/api_test/http_basic_login_test.rb test/integration/api_test/http_basic_login_with_api_token_test.rb test/integration/api_test/token_authentication_test.rb test/integration/application_test.rb test/integration/layout_test.rb test/integration/lib/redmine/menu_manager_test.rb test/integration/lib/redmine/themes_test.rb test/integration/routing_test.rb test/unit/activity_test.rb test/unit/attachment_test.rb test/unit/board_test.rb test/unit/category_test.rb test/unit/changeset_test.rb test/unit/comment_test.rb test/unit/custom_field_test.rb test/unit/custom_field_user_format_test.rb test/unit/custom_value_test.rb test/unit/default_data_test.rb test/unit/enabled_module_test.rb test/unit/enumeration_test.rb test/unit/group_test.rb test/unit/helpers/application_helper_test.rb test/unit/helpers/custom_fields_helper_test.rb test/unit/helpers/sort_helper_test.rb test/unit/helpers/timelog_helper_test.rb test/unit/issue_nested_set_test.rb test/unit/issue_priority_test.rb test/unit/journal_observer_test.rb test/unit/journal_test.rb test/unit/ldap_auth_source_test.rb test/unit/lib/open_project/database_test.rb test/unit/lib/redmine/access_control_test.rb test/unit/lib/redmine/ciphering_test.rb test/unit/lib/redmine/codeset_util_test.rb test/unit/lib/redmine/helpers/calendar_test.rb test/unit/lib/redmine/hook_test.rb test/unit/lib/redmine/i18n_test.rb test/unit/lib/redmine/menu_manager/mapper_test.rb test/unit/lib/redmine/menu_manager/menu_helper_test.rb test/unit/lib/redmine/menu_manager/menu_item_test.rb test/unit/lib/redmine/menu_manager_test.rb test/unit/lib/redmine/mime_type_test.rb test/unit/lib/redmine/notifiable_test.rb test/unit/lib/redmine/plugin_test.rb test/unit/lib/redmine/safe_attributes_test.rb test/unit/lib/redmine/scm/adapters/filesystem_adapter_test.rb test/unit/lib/redmine/scm/adapters/git_adapter_test.rb test/unit/lib/redmine/scm/adapters/subversion_adapter_test.rb test/unit/lib/redmine/unified_diff_test.rb test/unit/lib/redmine/wiki_formatting/macros_test.rb test/unit/lib/redmine/wiki_formatting/null_formatter_test.rb test/unit/lib/redmine/wiki_formatting/textile_formatter_test.rb test/unit/lib/redmine/wiki_formatting_test.rb test/unit/lib/redmine_test.rb test/unit/mail_handler_test.rb test/unit/member_test.rb test/unit/message_test.rb test/unit/principal_test.rb test/unit/project_nested_set_test.rb test/unit/project_test.rb test/unit/query_test.rb test/unit/relation_test.rb test/unit/repository_filesystem_test.rb test/unit/repository_git_test.rb test/unit/repository_subversion_test.rb test/unit/repository_test.rb test/unit/role_test.rb test/unit/search_test.rb test/unit/status_test.rb test/unit/time_entry_activity_test.rb test/unit/time_entry_test.rb test/unit/token_test.rb test/unit/type_test.rb test/unit/user_preference_test.rb test/unit/user_test.rb test/unit/version_test.rb test/unit/watcher_test.rb test/unit/wiki_content_test.rb test/unit/wiki_page_test.rb test/unit/wiki_redirect_test.rb test/unit/wiki_test.rb

* 180 conversions
    from: it { should ... }
      to: it { is_expected.to ... }

* 12 conversions
    from: obj.stub(:message)
      to: allow(obj).to receive(:message)

* 4 conversions
    from: Klass.any_instance.should_receive(:message)
      to: expect_any_instance_of(Klass).to receive(:message)

* 4 conversions
    from: obj.should_receive(:message)
      to: expect(obj).to receive(:message)

* 2 conversions
    from: Klass.any_instance.stub(:message)
      to: allow_any_instance_of(Klass).to receive(:message)

For more details: https://github.com/yujinakayama/transpec#supported-conversions

myabc committed:
Fix hash indenting in legacy routing spec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Prefer #map over #collect in legacy specs

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Customise spec Rake tasks to run legacy specs

For the time-being, we emulate Test::Unit Rake tasks and perform
separate runs for each of the legacy spec groups: there are still
problems with legacy specs not cleaning up after themselves.

Additionally legacy specs and the core spec suite cannot (yet) share
the same configuration.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove rails/test_helper require from legacy specs

This causes the minitest test runner to be loaded: this test runner
does not recognise the `--pattern` option and returns a non-zero exit
code.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Exclude legacy specs from regular 'core' spec run

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Split Travis spec build jobs into 'core', legacy

This restores the parallelization lost in dea3e576.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix repository fixture dirs used by scm:setup task

Broken in previous commit 2efd6f4e.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Adjust Attachment spec filesize assert for LF

On moving the fixture file, git appeared to change CRLF to LF.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Adjust Attachment spec digest assert

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove rspec/autorun require in legacy_spec_helper

We now only use the `rspec` or Rake commands.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix integration specs path in RSpec config

* Remove unnecessary includes.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Clear rspec-rails defined spec:legacy Rake task

This ensures legacy specs are not executed twice.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove functional spec rescue_action boilerplate

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove underscores from legacy example doc strings

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
https://community.openproject.org/work_packages/19483

myabc commented:
@kgalli @jonasheinrich on merging this, we should should adjust the Tests value in Jenkins config:
![config-core_config__jenkins_](https://cloud.githubusercontent.com/assets/755/7048190/bdc5b0e6-de11-11e4-8e5b-e0c14e878f5a.png)

jonasheinrich commented:
@myabc I'm watching the ticket now. When this gets updated to merged, I will adjust the jobs on our ci.

myabc committed:
Remove Test Unit section from RUNNING_TESTS doc

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
@jonasheinrich thanks

linki commented:
lgtm :+1:

myabc commented on .travis.yml:
```diff
@@ -49,13 +49,13 @@ env:
-    - "TEST_SUITE=test     DB=mysql2"
-    - "TEST_SUITE=spec     DB=mysql2"
+    - "TEST_SUITE=cucumber    DB=mysql2"
+    - "TEST_SUITE=spec:core   DB=mysql2"
```
Worth noting that although `TEST_SUITE=spec` would work fine here (albeit slowly, a run taking ~25 mins), I split the suites into two again in a5c3dace66ea90c3158002e8a2353f4da3b50f18 to take advantage of [parallelisation on Travis](http://docs.travis-ci.com/user/speeding-up-the-build/). This is something we might also want to consider for our internal CI.

It'd actually make more sense IMO to split up unit and feature specs, especially since the latter are more prone to flickering:

``` yaml
  - "TEST_SUITE=spec:models   DB=postgres"
  - "TEST_SUITE=spec:features DB=postgres"
```

However, this would warrant further discussion and is beyond the scope of this already huge PR.

/ @jonasheinrich @kgalli @NobodysNightmare @floriank

NobodysNightmare commented:
I am keen enough to throw this into the dev branch. That means `release/4.1` will be safe from any unforeseen consequences...

Except those that I did not foresee....
